gdnative-core/src/core_types/vector3.rs

```diff
@@ -10,65 +20,165 @@ pub enum Axis {
     ///
     /// This method runs faster than distance_to, so prefer it if you need to compare vectors or
     /// need the squared distance for some formula.
-    fn distance_squared_to(self, other: Vector3) -> f32;
+    #[inline]
+    pub fn distance_squared_to(self, other: Vector3) -> f32 {
+        other.glam().distance_squared(self.glam())
+    }
 
-    #[doc(hidden)]
-    #[allow(clippy::wrong_self_convention)]
-    fn to_sys(self) -> sys::godot_vector3;
-    /// Internal API for converting to `sys` representation. Makes it possible to remove
-    /// `transmute`s elsewhere.
-    #[doc(hidden)]
-    fn sys(&self) -> *const sys::godot_vector3;
-    /// Internal API for converting from `sys` representation. Makes it possible to remove
-    /// `transmute`s elsewhere.
-    #[doc(hidden)]
-    fn from_sys(v: sys::godot_vector3) -> Self;
-}
+    #[inline]
+    pub fn length(self) -> f32 {
+        self.glam().length()
+    }
+
+    #[inline]
+    pub fn normalize(self) -> Self {
+        Self::gd(self.glam().normalize())
+    }
 
-impl Vector3Godot for Vector3 {
     #[inline]
-    fn direction_to(self, other: Vector3) -> Vector3 {
-        (other - self).normalize()
+    pub fn dot(self, other: Self) -> f32 {
+        self.glam().dot(other.glam())
     }
 
     #[inline]
-    fn distance_to(self, other: Vector3) -> f32 {
-        (other - self).length()
+    pub fn cross(self, other: Self) -> Self {
+        Self::gd(self.glam().cross(other.glam()))
     }
 
     #[inline]
-    fn distance_squared_to(self, other: Vector3) -> f32 {
-        (other - self).square_length()
+    pub fn is_equal_approx(self, other: Self) -> bool {
+        relative_eq!(self.x, other.x)
+            && relative_eq!(self.y, other.y)
+            && relative_eq!(self.z, other.z)
     }
 
     #[inline]
-    fn to_sys(self) -> sys::godot_vector3 {
+    pub(crate) fn to_array(self) -> [f32; 3] {
+        [self.x, self.y, self.z]
+    }
+
+    #[doc(hidden)]
+    #[allow(clippy::wrong_self_convention)]
+    #[inline]
+    pub fn to_sys(self) -> sys::godot_vector3 {
         unsafe { std::mem::transmute(self) }
     }
 
+    /// Internal API for converting to `sys` representation. Makes it possible to remove
+    /// `transmute`s elsewhere.
+    #[doc(hidden)]
     #[inline]
-    fn sys(&self) -> *const sys::godot_vector3 {
+    pub fn sys(&self) -> *const sys::godot_vector3 {
         self as *const _ as *const _
     }
 
+    /// Internal API for converting from `sys` representation. Makes it possible to remove
+    /// `transmute`s elsewhere.
+    #[doc(hidden)]
     #[inline]
-    fn from_sys(v: sys::godot_vector3) -> Self {
+    pub fn from_sys(v: sys::godot_vector3) -> Self {
         unsafe { std::mem::transmute(v) }
     }
+
+    #[inline]
+    fn glam(self) -> Vec3A {
+        Vec3A::new(self.x, self.y, self.z)
+    }
+
+    #[inline]
+    fn gd(from: Vec3A) -> Self {
+        Self::new(from.x, from.y, from.z)
+    }
+}
+
+macro_rules! derive_op_impl {
```

Comment:
@toasteater would it make sense to extract these macros to their own private file `vector_ops.rs` and parametrize them over the vector type `Vector2`/`Vector3`?

Could avoid some code duplication and would keep the vector files cleaner.

Reply:
I think it makes sense, but I think it's ok to merge this as is for now and do on that later.

Reply:
OK for me 👍🏼